pkg/status/cel-evaluator.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types/ref"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// celEvaluator is a struct that holds the CEL environment
+// and provides a method to evaluate an expression with an unstructured object
+// as the context.
+type celEvaluator struct {
```

Comment:
You saw https://github.com/kubestellar/kubestellar/blame/v0.23.0-alpha.3/api/control/v1alpha1/types.go#L416-L418 , right? No need to start from scratch.

Reply:
My point is that Kubernetes already has code that supports using CEL with Kubernetes API objects, let's use it.

There are a lot of particulars, regarding the particular uses of CEL in Kubernetes, do not get distracted by that. But, if you are interested in the background, you might start by looking at how the type-checking is done for the CEL expressions that appear in `x-kubernetes-validations` items in the OpenAPI schema in a CustomResourceDefinition. https://github.com/kubernetes/apiextensions-apiserver/blob/v0.28.2/pkg/apis/apiextensions/validation/validation.go#L73 validates a CustomResourceDefinnition. That calls `validateCustomResourceDefinitionSpec`, which calls `validateCustomResourceDefinitionValidation` (to validate what the CRD says about how to validate the objects whose Kind that CRD is defining), which calls `ValidateCustomResourceDefinitionOpenAPISchema`, which calls a Compile method at https://github.com/kubernetes/apiextensions-apiserver/blob/v0.28.2/pkg/apis/apiextensions/validation/validation.go#L1112, which parses and type-checks (ultimately using cel-go's Env.Compile) all the CEL expressions in the OpenAPI schema. But let me emphasize again, this is more meta and complex than the problem that we have.

https://github.com/kubernetes/apiserver/blob/v0.28.2/pkg/cel/common/schemas.go#L40 is the heart of the generic part. This func takes an OpenAPI schema (which we can expect the resources processed by this feature to have) and produces a DeclType, which is the thing used in type-checking. The status controller could use this to type-check each CEL expression against each resource (kind of object) that it gets applied to. The parsing and type-checking would only have to be done once per (resource, CEL expression), not repeated for every version of every individual workload object's status.

Reply:
Will be addressed in followup PR - logged in issue #2141.